Athena driver

camsaul opened the pull request:
Started from PR #26241 but forked so we can make changes to it and get it green locally. The goal is for this to ship as part of 45.

Resolves #3960

### TODO

- [x] Create an s3 bucket to host the Athena driver; point Athena `deps.edn` towards this
- [x] Move namespaces added by the original PR like `metabase.driver.query-processor` to ones like `metabase.driver.athena.query-processor`
- [x] Copy driver test extensions from https://github.com/dacort/metabase-athena-driver/blob/master/test/metabase/test/data/athena.clj (not included in #26241 for whatever reason)
- [x] Fix Kondo errors, any other linter errors
    - [x] Add docstrings for everything or make it private.
- [x] Other misc code cleanup -- make code idiomatic, fix indentation, etc.
- [x] Set up an Athena test instance on AWS and get credentials
- [x] Get tests passing locally. Fix bugs
- [x] Only create tests databases if they don't exist already, or come up with some other way to prevent tests from stomping on each other if you do multiple test runs at the same time
- [x] Add credentials to CI and create a CI job for Athena 
- [x] Replace `println`s in the Athena test code with `log/`s
- [ ] Move the driver to a better S3 bucket? And strip out the Log4j config file while we're at it (@dpsutton)

### POLISH

I think it would be okay to do this stuff in follow-on PRs, I'm eager to get this in.

- [ ] Documentation in `docs/` etc.
- [ ] Re-enable nested Field support and get tests passing (see Damon's comments)
- [ ] Get in touch with Amazon Athena JDBC driver team and ask them nicely to host it on Maven Central or somewhere else normal
- [ ] Can you authenticate with any other method than AWS creds? If not, remove the "not recommended" from the DB details GUI (see comment below)

Polish is now being tracked in https://github.com/metabase/metabase/issues/27012


### Env Vars:

See also https://s3.amazonaws.com/athena-downloads/drivers/JDBC/SimbaAthenaJDBC-2.0.33.1002/docs/Simba+Amazon+Athena+JDBC+Connector+Install+and+Configuration+Guide.pdf

| Name | Notes |
| -- | -- |
| `MB_ATHENA_TEST_REGION` | I think this is something like `us-east-1` |
| `MB_ATHENA_TEST_ACCESS_KEY` | Creds for AWS user; needs perms for S3 bucket. See https://github.com/dacort/metabase-athena-driver#example-iam-policy |
| `MB_ATHENA_TEST_SECRET_KEY` | Creds for AWS user; needs perms for S3 bucket |
| `MB_ATHENA_TEST_S3_STAGING_DIR` | I believe this is supposed to be the S3 bucket in which to store query results, and the test data. Should be prefixed by `s3://`. Something like `s3://athena-driver-test-bucket/test/` seems to work, assuming you create that bucket and the AWS user has appropriate perms.  |

You can find these in 1password now.

### Note to Reviewers

If you see something that you don't like then please just push a commit to fix it to this branch rather than commenting or requesting changes. This PR is a collaborative effort, not Cam solo project! I have other stuff to squeeze in to 45 besides just polishing this all day

dacort committed:
Include Amazon Athena support for dacort/metabase-athena-driver

camsaul committed:
Use Metabase Maven repo to fetch Athena driver

camsaul committed:
Copy test extensions from Damon's Athena driver repo

camsaul committed:
Some code cleanup.

camsaul committed:
Move namespaces => metabase.driver.athena

camsaul committed:
Clean up the test extensions namespace

dacort commented:
FYI, there were a couple bugs I fixed when using the driver with v3 of the Athena engine. I hadn't had a chance to push them back up to the PR I opened, but you can find them here: https://github.com/dacort/metabase-athena-driver/commit/3dec67b50e70778ca7d98f9a1180e95a9be74245

And the closed issue for more context: https://github.com/dacort/metabase-athena-driver/issues/115

camsaul committed:
42 failures, 16 errors

camsaul committed:
Fix regex support; disable a few tests

camsaul committed:
Minor tweaks

camsaul committed:
Merge branch 'master' into dacort-athena-driver

camsaul committed:
Fix data-source-name

camsaul committed:
Fix :week and :day-of-week impls (mostly): 2 failures, 3 errors

camsaul committed:
Fix OFFSET, :week-of-year; skip test that has TIME column

camsaul committed:
Add Athena to CircleCI config. Don't wait for Java 11 tests to finish before driver tests.

camsaul committed:
ALL TESTS ARE PASSING! <3

camsaul committed:
Copy fixes for https://github.com/dacort/metabase-athena-driver/issues/115; add test

camsaul commented:
Thanks @dacort, I copied those changes and added a test as well to this PR.

camsaul committed:
We don't need to prep source files or fetch dependencies before running backend tests.

camsaul committed:
Fix Eastwood error.

camsaul committed:
Tweak CircleCI config.

camsaul commented on .circleci/config.yml:
```diff
@@ -589,16 +589,17 @@ workflows:
+              driver: ["bigquery-cloud-sdk", "googleanalytics", "redshift", "athena"]
           name: be-tests-<< matrix.driver >>-ee
           requires:
-            - be-tests-java-11-ee
```
Not really needed for this PR but by running all the driver tests as soon as we finish fetching deps (instead of waiting for regular Java 11 tests to finish) we should speed up CI a bit. The main reason I had them waiting for Java 11 tests in the first place was to avoid running a bunch of tests that will fail for sure and tying up our limited concurrency; that's not really an issue any more since we're on a different plan, so we should take advantage of that to get quicker feedback.

camsaul commented on .circleci/config.yml:
```diff
@@ -693,24 +694,17 @@ workflows:
                 when: always
           driver: presto-jdbc
 
-      - test-driver:
-          name: be-tests-redshift-ee
-          requires:
-            - be-tests-java-11-ee
-          driver: redshift
-          timeout: 15m
```
I moved this into the matrix with the other stuff above

camsaul commented on .github/workflows/backend.yml:
```diff
@@ -99,22 +99,9 @@ jobs:
     - name: Build static viz frontend
       run: yarn build-static-viz
 
-    - name: Compile Java source file(s)
-      run: clojure -X:deps prep
-    - name: Compile driver AOT namespaces
-      working-directory: modules/drivers
-      run: clojure -X:deps prep
-    - name: Fetch dependencies
-      run: clojure -P -X:dev:ci:ee:ee-dev:drivers:drivers-dev
-    - name: Fetch dependencies  (./bin/build/build-mb)
-      working-directory: bin/build-mb
-      run: clojure -P -M:test
-    - name: Fetch dependencies (./bin/build/build-drivers)
-      working-directory: bin/build-drivers
-      run: clojure -P -M:test
```
None of this stuff is needed before running backend unit tests, so not only does it potentially slow things down, it might also potentially break unrelated things e.g. unit tests can fail if there is an error in the build scripts `deps.edn`

camsaul commented on deps.edn:
```diff
@@ -160,6 +160,19 @@
+ ;; https://ask.clojure.org/index.php/12367/support-mvn-repos-inside-an-alias -- if we could, this could go in the
+ ;; `:drivers` alias instead.
+ :mvn/repos
+ {"athena"   {:url "https://s3.amazonaws.com/maven-athena"}
```
This is our s3 bucket, Dan said the name is a little confusing and I agreed. I think he was going to change it to something clearer. But that can be done in a follow-on PR if he doesn't get to it here

dpsutton replied:
I don't have the aws setup yet. I think its better to change it now before this ever goes live and anyone looks for artifacts in this bucket.
The idea is that the repo is ours, and might contain any number of artifacts. Naming it after one artifact in it doesn't make sense. Also, the artifact should not have metabase in its name. That is just a hosted jar and we don't want a classpath to end up with `metabase/athena-jdbc` and `amazon/athena-jdbc` at the same time but with the same (or worse, _similar_ but different) class files.

camsaul replied:
I agree with everything you're saying, but my main focus is on getting CI green right now. This PR is supposed to be a collaborative effort as opposed to a Cam solo project, so if you want to see that change then ping ops and get set up on AWS and change it? I have other stuff I need to try to squeeze in for SerDes before 45 so I really shouldn't be spending any more time polishing this

camsaul commented on deps.edn:
```diff
@@ -305,7 +318,7 @@
                                  "shared/src"
                                  "enterprise/backend/src"
-                                 "modules/drivers/bigquery/src"
+                                 "modules/drivers/athena/src"
```
We don't ship `:bigquery` (old version of the driver) anymore

camsaul commented on modules/drivers/athena/src/metabase/driver/athena/hive_parser.clj:
```diff
@@ -0,0 +1,62 @@
+(ns metabase.driver.athena.hive-parser
```
Don't ask me about this stuff because I didn't write it, Damon did. Unless there is something glaringly unacceptable I would vote towards getting this driver merged and doing cleanup as a follow-on if we need to

dacort replied:
@camsaul For future context struct/nested fields get returned as strings that we then have to parse in order to get the schema. So this hive parser recursively parses those strings into a schema. 

For example, this
```
array<struct<customerid:string,Order:bigint>>
```

Becomes this
```
[{"customerid": "string", "Order": "bigint"}]
```

camsaul replied:
Ok, we'll have to figure out how to get tests working for that. I turned off nested query support while getting the test suite green. I'll work on re-enabling them and making sure it's tested, maybe as a follow-on

camsaul committed:
Fix TIMESTAMP WITH TIME ZONE

camsaul committed:
Include the Athena/Redshift repos in the build-drivers deps.edn

camsaul committed:
Build and release scripts need to have the :mvn/repos as well

camsaul committed:
Merge branch 'master' into dacort-athena-driver

camsaul committed:
Revert change that enabled test for Presto

camsaul committed:
Un-enable failing test for Snowflake as well.

qnkhuat commented:
For anyone who's trying to run this and doesn't see Metabase print any logs, it's because the athena JDBC driver includes a `log4j2.properties` and log4j2 favor it over our log config.

To resolve I added `"-Dlog4j2.configurationFile=./resources/log4j2.xml"` to our `jvm-opts` in deps.

camsaul commented:
@qnkhuat oops I totally forgot to fix this upstream. `-Dlog4j2.configurationFile=./resources/log4j2.xml` would only fix it for `clj`, not for the uberjar. 

Locally I removed the `log4j2.xml` file from the Athena JDBC driver JAR in `~/.m2` manually. Since we're hosting the JAR ourselves we can do it there as well

camsaul commented:
One thing we might want to change is the admin UI for this... I don't know if there is any other way to connect besides using hard-coded IAM credentials? So we should probably stop saying it's "Not recommended" 

<img width="503" alt="image" src="https://user-images.githubusercontent.com/1455846/202530531-0d50acbf-4a0c-45b9-8f4f-bdb35780a020.png">

dacort commented:
> I don't know if there is any other way to connect besides using hard-coded IAM credentials? 

The recommended way to connect depends on where Metabase is running. If on an EC2 instance, [instance profiles](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-ec2_instance-profiles.html) can be used. If running on ECS, [IAM roles for tasks](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html) can be used. Both of these approaches result in the Athena driver automatically fetching session credentials based on whatever IAM role is configured.

Hard-coded credentials are typically not preferred because they must be stored somewhere (potentially unencrypted if db encryption is not enabled in Metabase), require manual work to rotate them on a regular basis, and if compromised can be used in perpetuity until they are disabled or rotated.

camsaul commented:
Thanks for the clarification @dacort 

We'll probably want to document that stuff. For our Cloud customers who can't easily control that stuff they'll probably need to continue using IAM credentials for the time being (maybe we can introduce some better way to do this in the future). So maybe we want to reword that a bit

dpsutton commented on modules/drivers/athena/src/metabase/driver/athena.clj:
```diff
@@ -0,0 +1,423 @@
+  [driver [_ arg pattern]]
+  (hsql/call :regexp_extract (sql.qp/->honeysql driver arg) pattern))
+
+;; keyword function converts database-type variable to a symbol, so we use symbols above to map the types
```
I don't know how to parse the sentence "keyword function converts database-type variable to a symbol"

dacort replied:
Had to poke around to figure this out myself. :) There's a list of mappings from Athena data types to Metabase data types up in `sql-jdbc.sync/database-type->base-type` that uses symbols as it's keys. The database type gets returned as a string so the `keyword` function converts the string database type to a symbol so it can then be used to lookup the proper Metabase type with that function. 😬 

I'm not entirely sure why I added the underlying mapping and didn't just leave the keys as strings and I apologize as I was learning clojure mid-implementation.

escherize committed:
moves all is clauses into the test

- previously only the first few tests were being run

escherize commented on modules/drivers/athena/src/metabase/driver/athena/hive_parser.clj:
```diff
@@ -0,0 +1,62 @@
+             (conj closes "]")
+             (str result "["))
+
+      (str/starts-with? schema "map<string,string>")
```
I noticed that non string->string maps for example `map<int,int>` won't parse properly. so that may be a problem. I tracked down some info on this schema format here: https://gist.github.com/kzhangkzhang/258d18858889fa97194011a249b74c43#complexcollection-types and it looks like maps can even have complex values as their keys.

I can look into fixing this piece.

dpsutton committed:
Sort ns in `metabase.driver.athena-test`

dpsutton committed:
Prevent athena's log4j2.properties file from becoming log config

athena includes log4j2.properties top-level with the properties:

```
status = debug
rootLogger.level=debug
```

And this kills our beautifully crafted logging. Set
"log4j2.configurationFile" in bootstrap to our own log4j2.xml.

Log4j2 looks in a few places for its logging config, the first of which
is the properties file. So when the jar is loaded, log4j2 considers this
a logging config change and we lose our logging.

dpsutton reviewed (approved):
I sorted a ns and added our log config file to bootstrap.clj so we don't have to worry about nuking the logger or jar exorcism.

I verified what happens if someone doesn't realize that we ship athena first party and they still include the dacort 3rd party jar in `plugins`. Things all seem to work. Wondering if we would want to log if we see the athena jar in the plugins directory perhaps?

camsaul committed:
Merge branch 'master' into dacort-athena-driver

camsaul committed:
!Merge branch 'master' into dacort-athena-driver

camsaul commented:
@salsakran I'll do that right now

camsaul commented:
@salsakran I created epic #27012 to track the polish stuff. Sorting out alternative auth mechanisms is probably the highest priority thing
